### ISSUE TITLE
результаты: тестирование на устойчивость, чувствительность к ошибкам

### ISSUE THREAD

**Issue by Amarchuk** [LerkaKostuik/Course-work#8]
Помимо остальных вещей, которые все надо делать, давайте сделаем тест на устойчивость к варьированию входных параметров. Это легче всего остального, но значимость имеет большую.

Делать так: для каждого или почти каждого входного параметра менять его значения по все карте на 5%, 10%, 15% и т.д., все остальные параметры не трогать. Получать двухжидкостное Qeff и в среднем понимать, на сколько процентов они отличаются по все карте от обычного значения Qeff что мы получили. Можно например взять среднее и 25%/75% квартили. И будет для каждого параметра график - как поменялся параметр по горизонтали и по вертикали как это отразилось на Q в среднем.

**Comment by ValerieKostiuk** [comment]
Я правильно понимаю, что для получения графика мне нужно будет азимутально усреднять Q после каждого варьирования?

**Comment by Amarchuk** [comment]
Нет, все проще: усреднять по всей карте. Т.е. у нас по сути будет карта в процентах как изменился Q и для нее найдем например среднее.

**Comment by ValerieKostiuk** [comment]
Я сейчас обновила версию своего кода, можете посмотреть, правильно ли я всё делаю. Нужно посмотреть файл error_sensitivity.ipynb.

**Comment by Amarchuk** [comment]
Да, это похоже. Вот только результаты-графики такие не бывают, поэтому вопросы:

1. А Вы не могли бы вот эту строчку объяснить:
`errors_min[i][j] = abs((Qeff_min_[i][j]*100.)/Qeff_min[i][j] - 100)`?

2. Лучше помимо линий на графиках рисовать еще и сами точки, если их немного

3. `np.sum(errors_min)/10075.` - у массивов в numpy можно сразу среднее брать, например `x.mean()`. И так же стандартное отклонение `x.std()`

4. давайте на графиках еще и ошибку какую-нибудь рисовать. Вот например тут же std

5. А работу с азимутальным усреднением мне тоже надо было тут посмотреть? Если да, то суть тут в следующем - можно делать двумя способами: усреднять исходные данные азимутально и уже для них находить Q, или же усреднять Q азимутально по двумерной карте. И в том и в другом случае получится одномерный график Q. Вы здесь делаете вторым способом, я в своей статье делал первым, можно и так и так, вопрос что мы хотим получить - смысл будет разный. Если воспроизвести результат Лероя - то первым.

**Comment by ValerieKostiuk** [comment]
> А Вы не могли бы вот эту строчку объяснить

На самом деле я здесь просто по-кривому определяю, на сколько процентов полученный после вариаций результат отличается от настоящего. (ну как в школе умножения крест на крест)

> А работу с азимутальным усреднением мне тоже надо было тут посмотреть? 

На самом деле, с азимутальным усреднением я работаю в Toomre.ipynb. Здесь я просто так баловалась и забыла убрать. Вы написали правильное замечание, я исправлю.

Может мне варьировать параметры с меньшим шагом, не 5 % . а например 3%?

> Лучше помимо линий на графиках рисовать еще и сами точки, если их немного

А какие точки рисовать, среднее значение изменения параметра Q в процентах в зависимости от изменения входного параметра?

**Comment by Amarchuk** [comment]
А, я понял, не так на скобочки посмотрел.

Меня в целом смущают графики, там по сути везде получается Q меняется линейно от входных параметров (причем еще и положительно везде, потому что `abs`). Кажется что выглядеть это все должно по-другому (проверьте пожалуйста что в коде нет какой-то явной ошибки). 

Давайте делать так: 
* Вы определяйте изменение со знаком, без abs (т.е. если было 1. и стало -1, то это -200%).
* для массива изменений Q будем рисовать 0.5 квартиль (медиана) как среднее изменение и к нему еще усы ошибок 0.25 и 0.75 квартили (это позволит учитывать как положительные, так и отрицательные изменения). Если непонятно что это такое - прочитайте, оно простое.
* можно рисовать используя `plt.boxplot`

Как всегда, если непонятно - попробуйте сделать одну-две попытки разобраться и потом с чистой совестью спрашивайте:)

**Comment by ValerieKostiuk** [comment]
Я построила boxplot для вариаций остальных данных  как вы написали (то есть убрала все лишние параметры) и увеличила пределы изменения. Также я загрузила все данные,которые я получила. 
На самом деле, я ожидала получить немного другие изображения, я думала, что медиана каждого box будет находится не в нуле и усы соответствующие минимальным и максимальным значениям будут другие (я долго пробовала их подвигать по вертикали). Также я думала, что box будут неодинаковыми для всех вариаций хотя бы для Sigma_s и sigma_HI (у них нелинейная зависимость).

**Comment by Amarchuk** [comment]
Это все очень странно. Точно

> plt.boxplot(array_errors,showfliers=False, positions=np.array(np.arange(-35.,40.,5.)),patch_artist = False, widths = 1.25) 

отрабатывает на массиве как надо? Попробуйте взять отдельно пару срезов и вычислить для них медиану. Как так может быть, что она одинаковая?

Еще может быть ошибка простая в коде. Например, если я правильно вижу, в цикле вы ставите `errors = []` только один раз, а потом в него только добавляете, а в самом цикле `array_errors.append(errors)`. Тут надо по логике обнулять список после добавления, чтобы на след. итерации в него не попадали предыдущие ошибки. Думаю в этом все дело.

А еще ответьте пожалуйста в #9.

**Comment by ValerieKostiuk** [comment]
Вы правы, я ошиблась в этом месте, спасибо, что указали мне на ошибку. Я исправила, стало гораздо лучше)

**Comment by Amarchuk** [comment]
Это один из навыков - пытаться осознать график. Придет со временем)

Да, теперь все лучше гораздо! Интересный график при Sigma_HI варьировании - усы обрезаются нулем. Хотите попробовать понять в чем тут дело?

**Comment by ValerieKostiuk** [comment]
И правда...я сначала не заметила,что получилось на этом графике. Конечно хотелось бы понять,почему так получается)

**Comment by Amarchuk** [comment]
Попробуйте подумать (ничего страшного если не выйдет, но не только же код писать). Возможно это связано как-то с тем, что нельзя сделать меньше Qstar. И вероятность ошибки тоже не стоит отбрасывать
